scenes/game_state/game_state.gd

```diff
@@ -3,5 +3,42 @@
 @tool
 extends Node
 
+## Emited when a new item is collected, even if it wasn't added to the
+## inventory due to it being already there.
+signal item_collected(item: InventoryItem)
+
+## Global inventory, used to track the items the player obtains and that
+## can be added to the loom.
 @export var inventory: Inventory = Inventory.new()
 @export var current_spawn_point: NodePath
+
+## Quest's items. Used to track the progress withing a story quest.
+var story_quest_inventory: Inventory = Inventory.new()
+
+
+## Resets the [member story_quest_inventory]. This needs to be called when
+## a new story quest starts.
+func quest_started() -> void:
+	story_quest_inventory.clear()
+
+
+## Adds the [InventoryItem] to the [member inventory] and the
+## [member story_quest_inventory], provided it wasn't already there.
+func item_was_collected(item: InventoryItem) -> void:
+	if not inventory.has_item(item):
+		inventory.add_item(item)
+	if not story_quest_inventory.has_item(item):
+		story_quest_inventory.add_item(item)
+	item_collected.emit(item)
+
+
+## Returns the items in the [member story_quest_inventory].[br]
+## Instead of returning a reference to the inventory or its internals, it
+## returns a new array.
+func items_collected_within_current_quest() -> Array[InventoryItem]:
+	return story_quest_inventory.items.duplicate()
```

Comment:
Can you explain why a duplicate copy of the items is needed?

Reply:
Yes!, it's just for encapsulation, so we don't expose the inventory's item array. If we return the array directly, and it gets mutated from where it was requested, the inventory will change (without any of the expected validations/signals, etc), if we return a copy, then the user can filter it or do whatever with that array and the inventory isn't affected.